precompiles/prototype/prototype_test.go

```diff
@@ -0,0 +1,339 @@
+
+	/*
+		Contract and ABI tests
+	*/
+
+	// Create a new IPrototypeContract instance and get Address and Abi.
+	contract := NewIPrototypeContract(k, appCodec, gasConfig)
+	require.NotNil(t, contract, "NewIPrototypeContract() should not return a nil contract")
+
+	address := contract.Address()
+	require.Equal(t, ContractAddress, address, "contract address should match the precompiled address")
+
+	abi := contract.Abi()
+	require.NotNil(t, abi, "contract ABI should not be nil")
+
+	// Check all methods are present in the ABI.
+	bech32ToHex := abi.Methods[Bech32ToHexAddrMethodName]
+	require.NotNil(t, bech32ToHex, "bech32ToHexAddr method should be present in the ABI")
+
+	bech32ify := abi.Methods[Bech32ifyMethodName]
+	require.NotNil(t, bech32ify, "bech32ify method should be present in the ABI")
+
+	getGasStabilityPoolBalance := abi.Methods[GetGasStabilityPoolBalanceName]
+	require.NotNil(t, getGasStabilityPoolBalance, "getGasStabilityPoolBalance method should be present in the ABI")
+
+	/*
+		Gas tests
+	*/
+
+	// Check all methods use the correct gas amount.
+	var method [4]byte
+
+	gasBech32ToHex := contract.RequiredGas(bech32ToHex.ID)
+	copy(method[:], bech32ToHex.ID[:4])
+	baseCost := uint64(len(method)) * gasConfig.WriteCostPerByte
+	require.Equal(
+		t,
+		GasRequiredByMethod[method]+baseCost,
+		gasBech32ToHex,
+		"bech32ToHexAddr method should require %d gas, got %d",
+		GasRequiredByMethod[method]+baseCost,
+		gasBech32ToHex)
+
+	gasBech32ify := contract.RequiredGas(bech32ify.ID)
+	copy(method[:], bech32ify.ID[:4])
+	baseCost = uint64(len(method)) * gasConfig.WriteCostPerByte
+	require.Equal(
+		t,
+		GasRequiredByMethod[method]+baseCost,
+		gasBech32ify,
+		"bech32ify method should require %d gas, got %d",
+		GasRequiredByMethod[method]+baseCost,
+		gasBech32ify)
+
+	gasGetGasStabilityPoolBalance := contract.RequiredGas(getGasStabilityPoolBalance.ID)
+	copy(method[:], getGasStabilityPoolBalance.ID[:4])
+	baseCost = uint64(len(method)) * gasConfig.WriteCostPerByte
+	require.Equal(
+		t,
+		GasRequiredByMethod[method]+baseCost,
+		gasGetGasStabilityPoolBalance,
+		"getGasStabilityPoolBalance method should require %d gas, got %d",
+		GasRequiredByMethod[method]+baseCost,
+		gasGetGasStabilityPoolBalance)
+
+	// Can not happen, but check the gas for an invalid method.
+	// At runtime if the method does not exist in the ABI, it returns an error.
+	invalidMethodBytes := []byte("invalidMethod")
+	gasInvalidMethod := contract.RequiredGas(invalidMethodBytes)
+	baseCost = uint64(len(method)) * gasConfig.WriteCostPerByte
+	require.Equal(
+		t,
+		uint64(0),
+		gasInvalidMethod,
+		"invalid method should require %d gas, got %d",
+		uint64(0),
+		gasInvalidMethod)
+}
+
+func Test_Bech32ToHexAddress(t *testing.T) {
+	/*
+		Configuration
+	*/
+
+	var encoding ethermint.EncodingConfig
+	appCodec := encoding.Codec
+	k, _, _, _ := keeper.FungibleKeeper(t)
+	gasConfig := storetypes.TransientGasConfig()
+
+	// Create contract and get ABI.
```

Comment:
Same as above consider breaking into separate tests under `Test_Bech32ToHexAddress` , same for other tests in this file